### PR TITLE
#2127 Search Improvements Help Panel

### LINKED ISSUE #2127: Search improvements. Help panel
As part of [ TBC: search improvements ticket ] user feedback suggested it is not always clear what fields are being searched and why some results are matching the searched term.

Here we propose changes to improve the search UX across all pages in Admin

**Search box**

**Help panel**

**'No results' message**

**To do**

- [ ] Search box. Add a label which identifies the data being search
- [ ] Search box. Include a Help link which, upon clicking it, opens a help panel
- [ ] Help panel. Ensure the help panel lists the fields being searched
- [ ] Help panel. Display a full-screen fade / overlay if possible. Alternatively show an 'in focus' highlight around the help panel.
- [ ] No results. Update the no results page to include the fields being searched
- [ ] No results. Update the no results page to include a 'Clear search & filters' button to empty the search & filter settings (if there are any)

### LINKED ISSUE #2128: Search improvements. No results message should identify fields being searched
As part of [Search Improvements](https://github.com/jac-uk/digital-platform/issues/875) we are proposing to tweak the text in the 'no results' message to include dynamic information about the fields available to search.

In addition add a 'Clear search' button after the 'no results' message. The button will only show if there is something to clear and the button label will reflect what is being cleared:
 - Clear filters
 - Clear search
 - Clear search & filters

---

**Current no results message**
image

---

**Proposed no results message**
image

---

**To do**

- [ ] Include dynamic information about the fields available to search
- [ ] Include a 'clear search' button

### PR DESCRIPTION
**Pull request by drieJAC**
## What's included?
Added help panel for search.
Updated empty search results panel.

Test at this url: https://jac-admin-develop--pr2139-feature-2127-search-ug76l1oz.web.app/

Closes #2127
Closes #2128 

## Who should test?
✅ Product owner
✅ Developers
✅ UTG

## How to test?
Ensure the following searches work correctly and display the correct information in the help box.

- Exercises Search
- Candidates Search
- Applications Search
- Independent Assessments Search
- Character Checks Search
- Stages (review)

See the screen recording below as an example of the Exercises Search.

https://github.com/jac-uk/admin/assets/115651787/bb033c83-57a7-4484-8715-407c2bde5087

## Risk - how likely is this to impact other areas?
🟠 Medium risk - this does change code that is shared with other areas

## Additional context
The Empty search should look like this:

266007148-3572f595-b89c-44de-983f-84d0b61f03af

---
PREVIEW:DEVELOP
_can be OFF, DEVELOP or STAGING_

### CODE REVIEW

File: src/mixins/searchMixin.js [hunk 1]
```diff
@@ -0,0 +1,13 @@
...
+    hasSearchMap() {
+      return this.searchMap !== null;
+    },
+    hasVisibleFields() {
+      return this.hasSearchMap && _has(this.searchMap, 'tooltip.visible') && this.searchMap.tooltip.visible.length;
+    },
+  },
+};
+
+export default searchMixin;
```
Comment (warrensearle) [inline src/mixins/searchMixin.js:13]: Is this being used anywhere in Admin?

**Review by warrensearle — CHANGES_REQUESTED** [review]

**Review by warrensearle — APPROVED** [review]
👍

**Review by HalcyonJAC — CHANGES_REQUESTED** [review]
I found a tiny issue. Please see the screen record. It seems that it will only happen when there is only one record in the table.

Example exercise: https://jac-admin-develop--pr2139-feature-2127-search-ug76l1oz.web.app/exercise/ZutEH7OHcjRlUIfXLyqP/applications/withdrawn

https://github.com/jac-uk/admin/assets/79906532/4718588c-b0ba-4c97-9233-ec61e7ba9b0e

**Review by tomlovesgithub — APPROVED** [review]

**Comment by drieJAC** [comment]
@warrensearle has said to ignore the small display bug that @HalcyonJAC reported above as in a future iteration we will modify the help panel so that it will appear above/below/left/right of the help button based on whether there is sufficient space on the screen. In doing so it will fix this bug.

**Pull request merged**